### PR TITLE
Gc stats

### PR DESCRIPTION
**Pull request by Ngoguey42**

### CODE REVIEW

**Comment by Ngoguey42** [comment]
`gc.ml` is abound 900 lines of code. I'd like to split it into `Gc_utils` (which would contain the new stats), `Gc_worker` and `Gc` at some point (not in this PR, after this PR is merged). @metanivek you did not fancy that solution last time we talked about this. WDYT now?

**Comment by metanivek** [comment]
I usually think that "utils" are a general code smell. Sometimes they are warranted as a collection of utility functions, but I usually want to pause and think about what I'm actually trying to model by stuffing things into a "utils". I don't know if this translates well, but I see them as the [junk drawer](https://en.wikipedia.org/wiki/Junk_drawer) of code bases. :wastebasket: 

As for gc specifically, I shied away from it previously because of the above utils issue, but also I didn't see a strong need to pull apart the module since it all works as one unit anyway (the almost identical `Args` module indicating this strongly to me). I'm also noticing that lots of OCaml code bases have large module files (sometimes in the thousands of lines), and I wonder if it is just more common to place more code into files with OCaml than other languages.

All that to say, I wouldn't be opposed to splitting the file, but I would just like to see a better modelling than a utils file. :wink:

**Comment by Ngoguey42** [comment]
This is now ready for review.

File: examples/gc.ml [hunk 1]
```diff
@@ -99,11 +99,19 @@ let rec megabytes_of_path path =
...
   let finished = function
-    | Ok (r : Store.Gc.stats) ->
+    | Ok (r : S.stats) ->
+        let steps = r.steps |> List.map (fun (k, v) -> (k, v.S.wall)) in
+        let duration = steps |> List.map snd |> List.fold_left Float.add 0. in
+        let finalisation_duration =
+          duration
+          -. List.assoc "worker startup" steps
+          -. List.assoc "before finalise" steps
+        in
```
Comment (metanivek) [inline examples/gc.ml:111]: I want to strongly advocate for keeping the simplified stats in the gc finished callback (and probably the entire stack of finalised stats passing). This level of complexity (and internal details knowledge needed) shouldn't be in the public api.

File: src/irmin-pack/unix/stats_intf.ml [hunk 2]
```diff
@@ -82,6 +84,109 @@ module File_manager = struct
   [@@deriving irmin]
 end
 
+module Latest_gc = struct
```
Comment (metanivek) [inline src/irmin-pack/unix/stats_intf.ml:87]: I think this naming can be improved. Could we name this `Gc` and then rename `t` to be `latest`? I think that `stats` is the `t` for this module.

File: src/irmin-pack/unix/stats_intf.ml [hunk 3]
```diff
@@ -93,15 +198,31 @@ module type Sigs = sig
...
   module File_manager : sig
-    include module type of File_manager
+    include module type of File_manager with type t = File_manager.t
+
+    type stat
+
+    val export : stat -> t
+  end
+
+  module Latest_gc : sig
```
Comment (metanivek) [inline src/irmin-pack/unix/stats_intf.ml:216]:
```suggestion
  module Gc : sig
```

File: src/irmin-pack/unix/mapping_file_intf.ml [hunk 4]
```diff
@@ -33,10 +33,14 @@ module type S = sig
   type open_error :=
     [ `Corrupted_mapping_file of string | `No_such_file_or_directory ]
 
+  type file_sizes := int63 * int63 * int63
+
   val create :
+    ?remember_file_sizes:(file_sizes -> unit) ->
```
Comment (metanivek) [inline src/irmin-pack/unix/mapping_file_intf.ml:39]:
```suggestion
    ?report_file_sizes:(file_sizes -> unit) ->
```

File: src/irmin-pack/unix/mapping_file.ml [hunk 5]
```diff
@@ -429,14 +430,20 @@ module Make (Io : Io.S) = struct
...
+    let* sorted_size = Io.size_of_path path1 in
+    Option.iter
+      (fun f -> f (reachable_size, sorted_size, mapping_size))
+      remember_file_sizes;
     Io.unlink path1 |> ignore;
 
     (* Open created map *)
-    open_map ~root ~generation
+    let+ mmap = open_map ~root ~generation in
+    mmap
```
Comment (metanivek) [inline src/irmin-pack/unix/mapping_file.ml:446]: Is this change needed?

File: src/irmin-pack/unix/io_errors.ml [hunk 6]
```diff
@@ -60,18 +60,26 @@ module Make (Io : Io.S) : S with module Io = Io = struct
...
-  let to_json_string result =
+  let json_string_of_result ok_t result =
     match result with
-    | Ok _ as v -> v |> Irmin.Type.to_json_string io_err_result
-    | Error e -> (
-        match e with
-        | `Io_misc _ as e -> Error e |> Irmin.Type.to_json_string io_err_result
-        | #error as e -> Error e |> Base.to_json_string)
+    | Ok _ as x ->
+        let typerepr = Irmin.Type.(result ok_t unreachable_t) in
```
Comment (metanivek) [inline src/irmin-pack/unix/io_errors.ml:69]: What is this witches brew?

File: src/irmin-pack/unix/gc.ml [hunk 7]
```diff
@@ -17,6 +17,310 @@
...
+        major_collections = v.major_collections;
+        heap_words = v.heap_words;
+        compactions = v.compactions;
+        top_heap_words = v.top_heap_words;
+        stack_size = v.stack_size;
+      }
+
+  let create : string -> t =
+   fun first_stepname ->
+    (* Reseting all irmin-pack stats. We'll reset again at every step. *)
```
Comment (metanivek) [inline src/irmin-pack/unix/gc.ml:187]: Maybe add a note that this is okay since we are in the worker process and therefore will not mess up the main process stats.

File: src/irmin-pack/unix/gc.ml [hunk 8]
```diff
@@ -17,6 +17,310 @@
...
+
+  type t = { stats : S.stats; timer : Steps_timer.t }
+  (** [t] is the running state while computing the stats *)
+
+  let create first_stepname ~commit_offset ~before_suffix_start_offset
+      ~before_suffix_end_offset ~after_suffix_start_offset =
+    let stats = Irmin.Type.(random S.stats_t |> unstage) () in
+    (* [repr] provides doesn't provide a generator that fills a type with
+       zeroes but it provides a random generator. Let's use it for our initial
+       value. *)
```
Comment (metanivek) [inline src/irmin-pack/unix/gc.ml:86]: This is convenient but want to flag that this will make it harder to debug in the future if something is not being captured correctly. I admit that the alternative of a manually constructed initial value isn't appealing given the size of the type.

File: src/irmin-pack/unix/gc.ml [hunk 9]
```diff
@@ -543,72 +881,72 @@ module Make (Args : Args) = struct
...
+                in
+                t.resulting_stats <- Some stats_builder;
 
                 [%log.debug
-                  "Gc ended. %a, newies bytes:%a" pp_stats !s Int63.pp
-                    (Int63.sub new_suffix_end_offset copy_end_offset)];
-                let () = Lwt.wakeup_later t.resolver (Ok !s) in
-                Ok (`Finalised !s)
+                  "Gc ended successfully. %a"
+                    (Irmin.Type.pp Stats.Latest_gc.stats_t)
```
Comment (metanivek) [inline src/irmin-pack/unix/gc.ml:946]: I'm concerned that this log line is a large output and too much information. Related to my plea to keep the simplified stats, I think it is better to log a minimal amount here (like we are currently doing).

**Review by metanivek — COMMENTED** [review]
I've done a high level review pass (not looking closely at the details of the stats calculations). I really (really) think we ought to separate these stats from the ones surfaced in the public API. Left a few other comments, but this my only real change request for this PR. Thanks for doing the work to get these stats in place!

File: src/irmin-pack/unix/gc.ml [hunk 10]
```diff
@@ -17,6 +17,310 @@
 open! Import
 include Gc_intf
 
+module Steps_timer = struct
```
Comment (icristescu) [inline src/irmin-pack/unix/gc.ml:20]: maybe these next three modules can be moved into a separate file `gc_stats.ml` for instance?

File: src/irmin-pack/unix/stats_intf.ml [hunk 11]
```diff
@@ -82,6 +84,109 @@ module File_manager = struct
   [@@deriving irmin]
 end
 
+module Latest_gc = struct
+  type rusage = {
```
Comment (icristescu) [inline src/irmin-pack/unix/stats_intf.ml:88]: In my mind there is a difference between stats that are specific to the code (like appended hashes) and the stats that can be recorded from outside (like rusage). Is there a way that the second type of stats could live in the bench/ package and register callbacks that update them? my concern is that the benchmark code gets too much intertwined with the "main" code, it will be harder to maintain both.  

Having a callback set to be the fun.id, would also make it faster when we are not benchmarking the code ? (even though it is probably fast enough to always record some data)

Lastly, this would avoid some code duplication between this files and the one in bench, right?

I'm not fixed on this, its just to have a discussion.

File: src/irmin-pack/unix/gc.ml [hunk 12]
```diff
@@ -17,6 +17,310 @@
...
+    let timer = Steps_timer.create first_stepname in
+    { stats; timer }
+
+  let finish_current_step t next_stepname =
+    let timer, prev_step = Steps_timer.progress t.timer next_stepname in
+    let stats = { t.stats with steps = prev_step :: t.stats.steps } in
+    { stats; timer }
+
+  let finalise t worker ~after_suffix_end_offset =
+    let t = finish_current_step t "will not appear in the stats" in
```
Comment (icristescu) [inline src/irmin-pack/unix/gc.ml:107]: I'm not sure I understand where you filter this step out

File: src/irmin-pack/unix/gc.ml (hunk 9 above)
Comment (metanivek) [reply src/irmin-pack/unix/gc.ml:949]: Didn't notice that it's debug. Agreed.

File: examples/gc.ml (hunk 1 above)
Comment (Ngoguey42) [reply examples/gc.ml:111]: I can tackle this by adding getters in `Stats.Latest_gc` that decipher the stats

Comment (metanivek) [reply examples/gc.ml:111]: I think that would work.

File: src/irmin-pack/unix/gc.ml (hunk 12 above)
Comment (Ngoguey42) [reply src/irmin-pack/unix/gc.ml:107]: `finalise` calls `finish_current_step` to finish the ongoing step. `finish_current_step` takes the name of the next step, there will be no next step

File: examples/gc.ml (hunk 1 above)
Comment (Ngoguey42) [reply examples/gc.ml]: Done. See latest commit

**Comment by Ngoguey42** [comment]
I'm done with this PR. I'll clean the commit history prior to merge, following approves

File: src/irmin-pack/unix/stats_intf.ml [hunk 13]
```diff
@@ -93,25 +193,55 @@ module type Sigs = sig
...
+        with type step = Latest_gc.step
+        with type worker = Latest_gc.worker
+        with type stats = Latest_gc.stats
+        with type t = Latest_gc.t
+
+    type stat
+
+    val export : stat -> t
+    val new_suffix_end_offset_before_finalise : worker -> int63
+    val finalise_duration : stats -> float
```
Comment (metanivek) [inline src/irmin-pack/unix/stats_intf.ml:226]: Can you add a doc to this? The most important being the unit of the time.

**Review by metanivek — CHANGES_REQUESTED** [review]
One small change requested and then you can `!fix!up!!`

File: src/irmin-pack/unix/stats_intf.ml (hunk 13 above)
Comment (Ngoguey42) [reply src/irmin-pack/unix/stats_intf.ml:227]: Done, and fixuped!
